### PR TITLE
[IOTDB-2500] [ query & cross-compaction ] cross-compaction stuck

### PR DESCRIPTION
**Pull request by choubenson**
After compaction, we have some operations as following:
(1) move tmp target files to final target files
(2) update memory, make it atomic.
(3) delete source files and its corresponding files

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java [hunk 1]
```diff
@@ -359,6 +359,22 @@ private static void updateOneTargetMods(
     targetFile.getModFile().close();
   }
 
+  public static void removeCompactionModification(
+      List<TsFileResource> selectedSeqTsFileResourceList,
+      List<TsFileResource> selectedUnSeqTsFileResourceList,
+      String fullStorageGroupName) {
+    try {
+      for (TsFileResource seqFile : selectedSeqTsFileResourceList) {
+        ModificationFile.getCompactionMods(seqFile).remove();
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java:368]: How about making getCompactionMods() as a method of TsFileResource

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java [hunk 2]
```diff
@@ -359,6 +359,22 @@ private static void updateOneTargetMods(
...
+      }
+      for (TsFileResource unseqFile : selectedUnSeqTsFileResourceList) {
+        ModificationFile.getCompactionMods(unseqFile).remove();
+      }
+    } catch (IOException e) {
+      logger.error("{} cannot remove merging modification ", fullStorageGroupName, e);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java:374]:
```suggestion
      logger.error("{} cannot remove compaction modification ", fullStorageGroupName, e);
```

Also print the mod file name

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/CrossSpaceCompactionExceptionHandler.java [hunk 3]
```diff
@@ -136,43 +127,77 @@ private static boolean handleWhenAllSourceFilesExist(
...
       throws IOException {
-    for (TsFileResource seqFile : seqFileList) {
-      ModificationFile compactionModFile = ModificationFile.getCompactionMods(seqFile);
-      if (compactionModFile.exists()) {
-        Collection<Modification> modifications = compactionModFile.getModifications();
-        ModificationFile normalModification = ModificationFile.getNormalMods(seqFile);
-        for (Modification modification : modifications) {
-          normalModification.write(modification);
+    TsFileResourceList unseqTsFileResourceList =
+        tsFileManager.getUnsequenceListByTimePartition(unseqFileList.get(0).getTimePartition());
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/CrossSpaceCompactionExceptionHandler.java:131]: How about passing the time partition in this method

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java [hunk 4]
```diff
@@ -172,18 +177,22 @@ private void executeCompaction()
   }
 
   private void updateTsFileResource() throws IOException {
+    seqTsFileResourceList.writeLock();
+    unseqTsFileResourceList.writeLock();
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java:181]: when we change the resourceList, we need to get the writelock of TsFileManager, not the two list. 
The query only get the read lock of TsFileManager

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java (hunk 2 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java:374]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java (hunk 4 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/CrossSpaceCompactionExceptionHandler.java [hunk 5]
```diff
@@ -136,43 +127,77 @@ private static boolean handleWhenAllSourceFilesExist(
...
+        tsFileManager.getSequenceListByTimePartition(seqFileList.get(0).getTimePartition());
+
+    // append new modifications to old mods files
+    CompactionUtils.appendNewModificationsToOldModsFile(seqFileList);
+    CompactionUtils.appendNewModificationsToOldModsFile(unseqFileList);
+
+    boolean removeAllTargetFile = true;
+    try {
+      seqTsFileResourceList.writeLock();
+      unseqTsFileResourceList.writeLock();
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/CrossSpaceCompactionExceptionHandler.java]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java [hunk 6]
```diff
@@ -171,6 +172,52 @@ public void addAll(List<TsFileResource> tsFileResourceList, boolean sequence) {
...
+      List<TsFileResource> seqFileResources,
+      List<TsFileResource> unseqFileResources,
+      List<TsFileResource> targetFileResources,
+      long timePartition)
+      throws IOException {
+    writeLock("replace");
+    try {
+      for (TsFileResource tsFileResource : seqFileResources) {
+        for (Map.Entry<Long, TsFileResourceList> entry : sequenceFiles.entrySet()) {
+          if (entry.getValue().contains(tsFileResource)) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java:186]: TsFileResourceList.remove() already check this, no need to check

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java [hunk 7]
```diff
@@ -171,6 +172,52 @@ public void addAll(List<TsFileResource> tsFileResourceList, boolean sequence) {
...
+      for (TsFileResource tsFileResource : unseqFileResources) {
+        for (Map.Entry<Long, TsFileResourceList> entry : unsequenceFiles.entrySet()) {
+          if (entry.getValue().contains(tsFileResource)) {
+            entry.getValue().remove(tsFileResource);
+            TsFileResourceManager.getInstance().removeTsFileResource(tsFileResource);
+            break;
+          }
+        }
+      }
+      if (unseqFileResources.size() == 0 || seqFileResources.size() != 0) {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java:202]: add a parameter in this method to indicate where to insert the targetTsFileResource

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java [hunk 8]
```diff
@@ -171,6 +172,52 @@ public void addAll(List<TsFileResource> tsFileResourceList, boolean sequence) {
...
+      List<TsFileResource> targetFileResources,
+      long timePartition)
+      throws IOException {
+    writeLock("replace");
+    try {
+      for (TsFileResource tsFileResource : seqFileResources) {
+        for (Map.Entry<Long, TsFileResourceList> entry : sequenceFiles.entrySet()) {
+          if (entry.getValue().contains(tsFileResource)) {
+            entry.getValue().remove(tsFileResource);
+            TsFileResourceManager.getInstance().removeTsFileResource(tsFileResource);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java:188]:
```suggestion
            if (entry.getValue().remove(tsFileResource)) {
                        TsFileResourceManager.getInstance().removeTsFileResource(tsFileResource);
            }
```

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/selector/RewriteCompactionFileSelector.java [hunk 9]
```diff
@@ -259,8 +259,12 @@ private void selectOverlappedSeqFiles(TsFileResource unseqFile) {
         long seqEndTime = seqFile.getEndTime(deviceId);
         long seqStartTime = seqFile.getStartTime(deviceId);
         if (unseqEndTime < seqStartTime) {
-          // if time range in unseq file is 10-20, seq file is 30-40, then skip this seq file and
-          // there is no more overlap later.
+          // if time range in unseq file is 10-20, seq file is 30-40, if this unseq file has no
+          // overlapped seq files, then select this seq file or skip this seq file and there is no
+          // more overlap later.
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/selector/RewriteCompactionFileSelector.java:264]:
```suggestion
          // Suppose the time range in unseq file is 10-20, seq file is 30-40, if this unseq file has no
          // overlapped seq files, then select this seq file and there is no
          // more overlap later.
```

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java [hunk 10]
```diff
@@ -172,18 +177,11 @@ private void executeCompaction()
   }
 
   private void updateTsFileResource() throws IOException {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java:179]: As there is only one caller, remove this method.

**Review by qiaojialin — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java (hunk 1 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java (hunk 6 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java (hunk 7 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/selector/RewriteCompactionFileSelector.java (hunk 9 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/selector/RewriteCompactionFileSelector.java]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java (hunk 8 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileManager.java]: resolved

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**